ipsec: New Prometheus metrics for XFRM configs

pchaigno opened the pull request:
Add new Prometheus metrics to report the number of keys, XFRM IN/OUT states, and XFRM IN/OUT/FWD policies. See commits for details.

pchaigno committed:
cmd, common: Move countUniqueIPsecKeys to common/ipsec pkg

Moving common IPsec helpers to their own package will allow us to call
them from both the agent and the Cilium CLIs.

Also move from checkmate to testing.T.

Note the prototype of countUniqueIPsecKeys changed to now take the XFRM
states in argument. That also allows us to make the corresponding test
unprivileged.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
ipsec: New Prometheus metric for number of keys in use

This new metric matches the existing information given by cilium encrypt
status, showing the number of IPsec keys currently in use on the node.
The number of keys should only change during key rotation and it should
go back to 1 after a configurable delay. This metric can help alert if
something went wrong during a key rotation.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
common/ipsec: Helper functions to count XFRM configs

Provide two new helper functions to count the number of XFRM policies
and states on the node, broken down by directions. XFRM policies can
have directions in, out, and fwd, whereas XFRM states can only be in or
out. For XFRM states, the direction is not a built-in parameter, but
something Cilium encodes in the packet mark.

These helper functions will be used in the subsequent commit to expose
new metrics.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

pchaigno committed:
ipsec: Expose XFRM config counts as metrics

This commit adds new Prometheus metrics for the number of XFRM states
and policies for each direction. These can be used to alert on XFRM
leaks. We had one such leak in the past for XFRM out policies which led
to a performance degradation when the node churn was high.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>

zacharysarah reviewed (approved):
@pchaigno Good work.

ti-mo commented on pkg/common/ipsec/utils_test.go:
Why add new tests using `checkmate`? We should make these `testing.T` instead.

pchaigno replied:
Because I didn't know about the switch.

gentoo-root commented on pkg/common/ipsec/utils_test.go:
```diff
@@ -27,6 +25,23 @@ func getXfrmState(src string, dst string, spi int, key string, mark uint32) netl
 			Key:    k,
 			ICVLen: 64,
 		},
+		Mark: &netlink.XfrmMark{
```
Nit: looks like it could be part of the first commit.

pchaigno replied:
It's not needed until this third commit, which is why I didn't include it in the first. If we were to revert the third and fourth commits later, I'd want this to be removed as well.

gentoo-root replied:
> If we were to revert the third and fourth commits later,

There would remain an unused parameter which could be misleading.

pchaigno replied:
Yep, exactly.

gentoo-root reviewed (approved):
Looks good to me.

chancez reviewed (approved):
It seems like these should be separate metrics rather than one metric with a bunch of different labels, but this seems to be the already established pattern for these metrics, so I'm not going to block on that, but I would encourage considering an approach using separate metrics. 

Reasoning: usually you use labels for different dimensions of a given metric, and you can aggregate over the dimensions in a meaningful way. Given that each of these labels represents different "statistics", you can't really meaningfully aggregate them, which is why I suggest that they are broken apart into separate metrics instead.

Again, non-blocking because your following existing conventions, but it might be good to create a follow up to revisit these xfrm metrics as a whole.

tommyp1ckles commented on pkg/datapath/linux/ipsec/xfrm_collector.go:
```diff
@@ -101,5 +124,19 @@ func (x *xfrmCollector) Collect(ch chan<- prometheus.Metric) {
+func (x *xfrmCollector) collectConfigStats(ch chan<- prometheus.Metric) {
+	states, err := x.xfrmStateFunc()
+	if err != nil {
+		log.WithError(err).Error("Error while getting XFRM states")
```
nit: would it better to return the error here and log at the collect level with some context of what the outcome of this error is.

tommyp1ckles commented on pkg/datapath/linux/ipsec/xfrm_collector.go:
```diff
@@ -134,6 +138,20 @@ func (x *xfrmCollector) collectConfigStats(ch chan<- prometheus.Metric) {
+
+	policies, err := x.xfrmPolicyFunc()
+	if err != nil {
+		log.WithError(err).Error("Error while getting XFRM policies")
```
Same question here, would be nice to know that this only affect metrics

tommyp1ckles reviewed (approved):
Changes look good, sorry for the delay - was on PTO.

pchaigno committed:
ipsec: Remove unit tests for xfrm_collector

Remove the unit tests for xfrm_collector. They are simply not worth it.
All xfrm_collector does is collect statistics via helper functions and
feed them into Prometheus functions. The helper functions are covered by
unit tests. Prometheus functions are covered by unit tests. We're really
just passing the information from one to the other.

But to cover this trivial logic in unit tests, we end up copying lots of
information across the code and tests. We also need to pass the helper
functions as a function pointer to xfrm_collector. Every new metric is
going to need a new function pointer.

I believe this is a waste of engineering time and very unlikely to catch
mistakes. It doesn't even test the helper functions since we mock them
out. So it's really unclear what we're trying to cover here. Removing it
makes it easier to add new metrics in followup commits.

Signed-off-by: Paul Chaignon <paul.chaignon@gmail.com>
